File: backend/models/foodLog.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

// This small schema will store the key nutrients
const nutrientSchema = new Schema({
    calories: { type: Number, default: 0 },
    protein: { type: Number, default: 0 },
    fat: { type: Number, default: 0 },
    carbohydrates: { type: Number, default: 0 }
}, { _id: false });

const foodLogSchema = new Schema({
    user: {
        type: Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    // The date the food was eaten
    date: {
        type: Date,
        default: Date.now,
        required: true
    },
    // What meal this was for (Breakfast, Lunch, etc.)
    mealType: {
        type: String,
        enum: ['Breakfast', 'Lunch', 'Dinner', 'Snack'],
        required: true
    },
    // Details from the scanned/searched product
    product_name: {
        type: String,
        required: true
    },
    brands: {
        type: String
    },
    image_url: {
        type: String
    },
    // The nutritional info for the item
    nutrients: nutrientSchema,

    // You could also add fields like 'servingSize' or 'quantity' here

}, {
    timestamps: true // Adds createdAt and updatedAt
});

// Indexing will make it much faster to fetch logs for a user on a specific date
foodLogSchema.index({ user: 1, date: -1 });

module.exports = mongoose.model('FoodLog', foodLogSchema);

File: fyp/lib/models/food_log.dart
// lib/models/food_log.dart

import 'dart:convert';

class FoodLog {
  final String id;
  final String user;
  final DateTime date;
  final String mealType;
  final String productName;
  final String? brands;
  final String? imageUrl;
  final Nutrients nutrients;

  FoodLog({
    required this.id,
    required this.user,
    required this.date,
    required this.mealType,
    required this.productName,
    this.brands,
    this.imageUrl,
    required this.nutrients,
  });

  factory FoodLog.fromJson(Map<String, dynamic> json) {
    return FoodLog(
      // Use '_id' from MongoDB
      id: json['_id'] as String, 
      user: json['user'] as String,
      date: DateTime.parse(json['date'] as String),
      mealType: json['mealType'] as String,
      productName: json['product_name'] as String,
      brands: json['brands'] as String?,
      imageUrl: json['image_url'] as String?,
      nutrients: Nutrients.fromJson(json['nutrients'] as Map<String, dynamic>),
    );
  }

  static List<FoodLog> fromJsonList(List<dynamic> jsonList) {
    return jsonList.map((json) => FoodLog.fromJson(json as Map<String, dynamic>)).toList();
  }
}

class Nutrients {
  final double calories;
  final double protein;
  final double fat;
  final double carbohydrates;

  Nutrients({
    required this.calories,
    required this.protein,
    required this.fat,
    required this.carbohydrates,
  });

  factory Nutrients.fromJson(Map<String, dynamic> json) {
    return Nutrients(
      calories: (json['calories'] ?? 0).toDouble(),
      protein: (json['protein'] ?? 0).toDouble(),
      fat: (json['fat'] ?? 0).toDouble(),
      carbohydrates: (json['carbohydrates'] ?? 0).toDouble(),
    );
  }
}